### DISCUSSION TITLE
Recommended way to check the database is working

### DISCUSSION THREAD

**Discussion by kisamoto** [pocketbase/pocketbase#5035] [Q&A] [answered]
I'd like to extend the `/api/health` endpoint to do more than just return if the server is up.

This will include checking the database is accessible.

Is there a recommended way to achieve this?

I know there is the [`sql.Ping`](https://pkg.go.dev/database/sql#DB.Ping) however as the `.DB()` method is marked deprecated and using `ConcurrentDB()` only returns a `Builder` which does not expose `Ping` or the underlying `*sql.DB` I was wondering if there was another way.

**Comment by ganigeorgiev** [comment, marked as answer]
> I'd like to extend the /api/health endpoint to do more than just return if the server is up.

Why? What is the problem you are experiencing and trying to solve?

If the `/api/health` is accessible it technically means that the `app.Bootstrap()` has been executed successfully, aka. the db connection was initialized successfully, so there shouldn't be need to check this manually.

**Reply by kisamoto** [reply]
Will this only check upon app start but not if something happens while the app is running?

I'm look to monitor and safeguard against things like:

- Database locked;
- Disk IOPS maxed out;
- Storage connection failure (using network backed storage);
- Storage lock (I mount the DB via docker volume and also into a second container which dumps the SQLite DB);

I do also have Grafana monitoring underlying infrastructure but this would be an additional endpoint to help monitor API uptime and possibly (in future - not a use case now) a load balancer could switch to a new instance.

**Reply by ganigeorgiev** [reply]
I'm still not sure that I understand your concerns and I think you are trying to solve a problem that doesn't really exist.

The Dao queries have auto fail-retry by default in addition to the SQLite `busy_timeout`.
Even in the rare case case of db locks, what is your plan for resolution/reaction for this event? 

You can "ping" the database on each health check (_the `dbx.Builder` is interface and you can type assert it if you want to access its underlying `sql.DB`_) but I consider this wasteful and unnecessary (_and keep in mind that the health endpoint is public_).

**Reply by kisamoto** [reply]
I'd like to be proactively alerted more than anything else. App doesn't have a lot of traffic at the moment so if something does happen there's little chance I'd be aware.

Also to expand, I'm going to also check external network connectivity in the health endpoint.

I had a network configuration issue on my VPS which blocked DNS resolution for external networks. I'd like to add that to my new health endpoint. Although I know how to do that - not related to the database check.

**Reply by kisamoto** [reply]
So I think I'll just use a `SELECT id LIMIT 1` type query for now. It will check access (and using a Dao query will enable some retries) but not writeability. Good enough compromise for me for now.

Thanks for highlighting the concerns about type assertion and pinging on healthcheck. FYI I've added HTTP Basic Auth and rate limiting to the endpoint (all being well it should only be hit once every second so I don't think this is too bad in the scheme of things).

Appreciate you taking the time to reply and happy hacking.

**Reply by TobiasNickel** [reply]
what about just making a table (naming it let's say:,... healthcheck.) with a single entry, You can update it and see if it succeed. I am also the opinion, that pocketbase don't need the extra overhead.

**Reply by kisamoto** [reply]
Yeah not a bad idea - thanks.

Appreciate that you guys don't think it needs the extra overhead, that's fine and I'm not suggesting contributing this to the core library at all.

But personally I'd like to have the peace of mind of going beyond 'is the app running'.
